middleware/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,213 @@
+// Package kubernetes provides the kubernetes backend.
+package kubernetes
+
+import (
+    "fmt"
+	"strings"
+	"time"
+
+	"github.com/miekg/coredns/middleware"
+	"github.com/miekg/coredns/middleware/kubernetes/msg"
+	k8sc "github.com/miekg/coredns/middleware/kubernetes/k8sclient"
+	"github.com/miekg/coredns/middleware/proxy"
+//	"github.com/miekg/coredns/middleware/singleflight"
+
+    "github.com/miekg/dns"
+	"golang.org/x/net/context"
+)
+
+type Kubernetes struct {
+	Next       middleware.Handler
+	Zones      []string
+	Proxy      proxy.Proxy // Proxy for looking up names during the resolution process
+	Ctx        context.Context
+//	Inflight   *singleflight.Group
+    APIConn    *k8sc.K8sConnector
+}
+
+
+func (g Kubernetes) getZoneForName(name string) (string, []string) {
+    var zone string
+    var segments []string
+
+    for _, z := range g.Zones {
+        if dns.IsSubDomain(z, name) {
+            zone = z 
+    
+            segments = dns.SplitDomainName(name)
+            segments = segments[:len(segments) - dns.CountLabel(zone)]
+        }
+    }   
+
+    return zone, segments
+} 
+
+
+// Records looks up services in kubernetes.
+// If exact is true, it will lookup just
+// this name. This is used when find matches when completing SRV lookups
+// for instance.
+func (g Kubernetes) Records(name string, exact bool) ([]msg.Service, error) {
+
+    zone, serviceSegments := g.getZoneForName(name)
+
+    var serviceName string
+    var namespace string
+
+    // For initial implementation, assume namespace is first serviceSegment
+    // and service name is remaining segments.
+    serviceSegLen := len(serviceSegments)
+    if serviceSegLen >= 2 {
+        namespace = serviceSegments[serviceSegLen-1]
+        serviceName = strings.Join(serviceSegments[:serviceSegLen-1], ".")
+    }
+    // else we are looking up the zone. So handle the NS, SOA records etc.
+
+    fmt.Println("[debug] zone: ", zone)
+    fmt.Println("[debug] servicename: ", serviceName)
+    fmt.Println("[debug] namespace: ", namespace)
+    fmt.Println("[debug] APIconn: ", g.APIConn)
+
+    k8sItem := g.APIConn.GetServiceItemInNamespace(namespace, serviceName)
+    fmt.Println("[debug] k8s item:", k8sItem)
+
+    switch {
+        case exact && k8sItem == nil:
+            fmt.Println("here2")
+            return nil, nil
+    }
+
+    if k8sItem == nil {
+        // Did not find item in k8s
+        return nil, nil
+    }
+
+    fmt.Println("[debug] clusterIP:", k8sItem.Spec.ClusterIP)
```

Comment:
I'm open to suggestions. At the moment, I don't fully understand what semantics this Records() function needs to provide. I'm continuing to debug the etcd middleware to get an understanding. Do you have anything written that describes how this part of coredns works?

For now, leaving this as is.